src/history/hs.c

```diff
@@ -746,45 +746,45 @@ int
 __wt_hs_cursor_position(WT_SESSION_IMPL *session, WT_CURSOR *cursor, uint32_t btree_id,
   WT_ITEM *key, wt_timestamp_t timestamp)
 {
-    WT_ITEM hs_key;
-    WT_TIME_PAIR hs_start, hs_stop;
-    uint32_t hs_btree_id;
+    WT_DECL_ITEM(srch_key);
+    WT_DECL_RET;
     int cmp, exact;
+    bool set_key;
+
+    set_key = false;
+
+    WT_RET(__wt_scr_alloc(session, 0, &srch_key));
 
     /*
      * Because of the special visibility rules for the history store, a new key can appear in
      * between our search and the set of updates that we're interested in. Keep trying until we find
      * it.
+     *
+     * Note that we need to compare the raw key off the cursor to determine where we are in the
+     * history store as opposed to comparing the embedded data store key since the ordering is not
+     * guaranteed to be the same.
      */
     for (;;) {
         cursor->set_key(cursor, btree_id, key, timestamp, WT_TXN_MAX, WT_TS_MAX, WT_TXN_MAX);
-        WT_RET(cursor->search_near(cursor, &exact));
+        if (!set_key) {
```

Comment:
I believe you are only populating the `srch_key` once in the `for` loop. I would suggest moving out the call for `__wt_buf_set()` from the `for` loop and getting rid of the variable `bool set_key`.

Reply:
I have to do it after the `set_key` though. And I believe I need to do the `set_key` on each iteration before I call `search_near`.

Reply:
Ah I get your point now. You only need to do it once and only after first call to `set_key()`. Then I assume there is no better way to achieve it.

Reply:
2654a11

Left a comment to make it clearer why I'm doing this.